File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/binary/Metadata.kt
package com.tyhp.lang.binary

import com.tyhp.lang.settings.InstallMode
import com.tyhp.lang.settings.parseInstallMode
import java.nio.file.Files
import java.nio.file.Path

const val PLUGIN_INSTALLER_ID = "com.tyhp.lang"

data class InstallMetadata(
    val installedBy: String,
    val version: String,
    val mode: InstallMode,
    val assetName: String,
    val installedAt: String,
)

fun metadataPath(cliDir: Path): Path = cliDir.resolve("metadata.json")

fun readInstallMetadata(cliDir: Path): InstallMetadata? {
    val file = metadataPath(cliDir)
    return try {
        if (!Files.isRegularFile(file)) {
            return null
        }
        parseInstallMetadata(Files.readString(file))
    } catch (_: Exception) {
        null
    }
}

fun writeInstallMetadata(cliDir: Path, metadata: InstallMetadata) {
    Files.createDirectories(cliDir)
    Files.writeString(metadataPath(cliDir), "${toJson(metadata)}\n")
}

fun deleteInstallMetadata(cliDir: Path) {
    try {
        Files.deleteIfExists(metadataPath(cliDir))
    } catch (_: Exception) {
        // ignore
    }
}

fun isPluginOwnedInstall(metadata: InstallMetadata?): Boolean =
    metadata != null &&
        metadata.installedBy == PLUGIN_INSTALLER_ID &&
        metadata.mode == InstallMode.EXTENSION

fun parseInstallMetadata(json: String): InstallMetadata? {
    val parsed = parseJson(json) as? Map<*, *> ?: return null
    val installedBy = parsed["installedBy"] as? String ?: return null
    val version = parsed["version"] as? String ?: return null
    val mode = parseInstallMode(parsed["mode"] as? String)
    val assetName = parsed["assetName"] as? String ?: ""
    val installedAt = parsed["installedAt"] as? String ?: ""
    return InstallMetadata(
        installedBy = installedBy,
        version = version,
        mode = mode,
        assetName = assetName,
        installedAt = installedAt,
    )
}

fun toJson(metadata: InstallMetadata): String {
    return buildString {
        append("{\n")
        append("  \"installedBy\": ${jsonQuote(metadata.installedBy)},\n")
        append("  \"version\": ${jsonQuote(metadata.version)},\n")
        append("  \"mode\": ${jsonQuote(metadata.mode.value)},\n")
        append("  \"assetName\": ${jsonQuote(metadata.assetName)},\n")
        append("  \"installedAt\": ${jsonQuote(metadata.installedAt)}\n")
        append("}")
    }
}

private fun jsonQuote(value: String): String {
    val escaped = value
        .replace("\\", "\\\\")
        .replace("\"", "\\\"")
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
    return "\"$escaped\""
}

File: tyhp-lang/vscode/src/binary/metadata.ts
import * as fs from "fs";
import * as path from "path";
import { InstallMode } from "../config/settingsCore";

export const EXTENSION_ID = "tyhp-lang.tyhp";

export interface InstallMetadata {
    installedBy: string;
    version: string;
    mode: InstallMode;
    assetName: string;
    installedAt: string;
}

export function metadataPath(cliDir: string): string {
    return path.join(cliDir, "metadata.json");
}

export function readInstallMetadata(cliDir: string): InstallMetadata | undefined {
    const file = metadataPath(cliDir);
    try {
        if (!fs.existsSync(file)) {
            return undefined;
        }
        const parsed = JSON.parse(fs.readFileSync(file, "utf8")) as InstallMetadata;
        if (!parsed || typeof parsed !== "object") {
            return undefined;
        }
        return parsed;
    } catch {
        return undefined;
    }
}

export function writeInstallMetadata(cliDir: string, metadata: InstallMetadata): void {
    fs.mkdirSync(cliDir, { recursive: true });
    fs.writeFileSync(metadataPath(cliDir), `${JSON.stringify(metadata, null, 2)}\n`, "utf8");
}

export function deleteInstallMetadata(cliDir: string): void {
    const file = metadataPath(cliDir);
    try {
        if (fs.existsSync(file)) {
            fs.unlinkSync(file);
        }
    } catch {
        // ignore
    }
}

export function isExtensionOwnedInstall(metadata: InstallMetadata | undefined): boolean {
    return (
        metadata !== undefined &&
        metadata.installedBy === EXTENSION_ID &&
        metadata.mode === "extension"
    );
}
